src/gasPriceOracle/oracle.ts

```diff
@@ -0,0 +1,98 @@
+import { BigNumber, providers } from "ethers";
+import { toBN } from "../utils";
+
+export type GasPriceEstimate = {
+  maxFeePerGas: BigNumber;
+  maxPriorityFeePerGas?: BigNumber;
+};
+
+type ValidatedFeeData = {
+  [idx: string]: BigNumber;
+  gasPrice: BigNumber;
+  maxFeePerGas: BigNumber;
+  maxPriorityFeePerGas: BigNumber;
+};
+
+interface GasPriceFeed {
+  (provider: providers.Provider): Promise<GasPriceEstimate>;
+}
+
+async function getProviderFeeData(provider: providers.Provider): Promise<ValidatedFeeData> {
+  // Start with safe defaults.
+  const feeData: ValidatedFeeData = {
+    gasPrice: toBN(Number.MAX_SAFE_INTEGER),
+    maxFeePerGas: toBN(Number.MAX_SAFE_INTEGER),
+    maxPriorityFeePerGas: toBN(Number.MAX_SAFE_INTEGER),
+    // lastBaseFeePerGas: toBN(Number.MAX_SAFE_INTEGER), @todo: ethers 5.7.x
+  };
+
+  try {
+    const response: providers.FeeData = await provider.getFeeData();
+    const expectedKeys: string[] = Object.keys(feeData);
+    Object.entries(response).forEach(([key, value]) => {
+      if (expectedKeys.includes(key) && BigNumber.isBigNumber(value) && value.gt(0)) {
+        feeData[key] = toBN(value);
```

Comment:
Isn't this already a BigNumber?

Reply:
I _thought_ I needed to pipe it through toBN() because the FeeData definition permits all members to be null. Reflecting on that, it was obviously wrong because toBN() wouldn't accept a null anyway.

In any case, the logic on line 33 guarantees that it is a BigNumber, so I can actually remove toBN() here. Nice!

```
export interface FeeData {
    lastBaseFeePerGas: null | BigNumber;
    maxFeePerGas: null | BigNumber;
    maxPriorityFeePerGas: null | BigNumber;
    gasPrice: null | BigNumber;
}
```